### PR TITLE
feat(inductor): Add `RAdam` to Inductor by converting data-dependent control-flow to `torch.where`

### PR DESCRIPTION
**Pull request by jon-chuang**
For small epochs (adaptive learning rate = 1), this will be more costly than not computing `rect` and `adaptive_lr`, but unlikely by much - they are all fused into a single kernel anyway.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by vadimkantorov** [comment]
I hope at some point, we can have polymorphic torch.sqrt(1.0) -> 1.0 for python scalars just like in NumPy and torch.pi to not have to import math just for these simple things :) or speed up https://github.com/jon-chuang/pytorch/blob/89eb7a75a251c41c4bee86e9ede1001b0d3998af/torch/optim/optimizer.py#L84 cpu scalar tensors

File: torch/optim/radam.py [hunk 1]
```diff
@@ -379,22 +378,23 @@ def _multi_tensor_radam(
...
-            _dispatch_sqrt(
-                (rho_t - 4)
-                * (rho_t - 2)
-                * rho_inf
-                / ((rho_inf - 4) * (rho_inf - 2) * rho_t)
-            )
-            if rho_t > 5
-            else 0
-            for rho_t in rho_t_list
+            torch.where(
```
Comment (jon-chuang) [inline torch/optim/radam.py:381]: Hmm, not sure this is a good idea in the eager case? Anyway, it executes on CPU, but still might be a bit slow

File: torch/optim/radam.py [hunk 2]
```diff
@@ -379,22 +378,23 @@ def _multi_tensor_radam(
...
         ]
-        unrectified = [0 if rect > 0 else 1.0 for rect in rect]
+        unrectified = [torch.where(r > 0, 0.0, 1.0) for r in rect]
 
         bias_correction1 = [1 - beta1 ** _get_value(step) for step in grouped_state_steps]
-        unrect_step_size = _stack_if_compiling([(lr * rect / bc) * -1 for rect, bc in zip(unrectified, bias_correction1)])
+        unrect_step_size = [(lr * _get_value(rect) / bc) * -1 for rect, bc in zip(unrectified, bias_correction1)]
         bias_correction2_sqrt_times_rect_step_size = [
-            _dispatch_sqrt(1 - beta2 ** _get_value(step)) * (lr * rect / bc) * -1
+            _dispatch_sqrt(1 - beta2 ** _get_value(step)) * (lr * _get_value(rect) / bc) * -1
```
Comment (jon-chuang) [inline torch/optim/radam.py:397]: We should probably have a "capturable" code path similar to Adam/NAdam and now Adagrad (https://github.com/pytorch/pytorch/pull/110339) / Adamax

It can use `foreach_sqrt` for the GPU path? :thinking: 

The only reason we wouldn't use it here is because we want this bias correction to be on the Scalar codepath, but that doesn't really make that much sense since it is consumed in foreach below.

File: torch/optim/optimizer.py [hunk 3]
```diff
@@ -100,6 +100,12 @@ def _dispatch_sqrt(x: float):  # float annotation is needed because of torchscri
     else:
         return math.sqrt(x)
 
+def _dispatch_abs(x: float):  # float annotation is needed because of torchscript type inference
+    if not torch.jit.is_scripting() and isinstance(x, torch.Tensor):
+        return x.abs()
+    else:
+        return abs(x)
```
Comment (peterbell10) [inline torch/optim/optimizer.py:107]: I improved dynamo's `abs` support in #110398 so this function shouldn't be needed any more.

File: torch/optim/radam.py (hunk 2 above)
Comment (jon-chuang) [reply torch/optim/radam.py:397]: hmm, I think the issue with converting to foreach here is the `_get_value` part, but honestly I could do list comprehension to create the corresponding step values.

File: torch/optim/radam.py (hunk 1 above)
Comment (janeyx99) [reply torch/optim/radam.py:381]: Putting this on the capturable path would help resolve this as well

File: torch/optim/radam.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/radam.py:397]: Adding a capturable code path may be the right move here. Then, we could get rid of the need for get_value as the capturable path should use the foreach ops, and the eager path can still keep scalar math on python.

**Review by janeyx99 — COMMENTED** [review]
Agreed that a capturable path will more sense here. However, adding capturable well requires more work as you should also support it in the single tensor case and add the right test cases in test_cuda and test_optim.

**Comment by vadimkantorov** [comment]
> I hope at some point, we can have polymorphic torch.sqrt

For this specific case, can torch.sqrt call be replaced by `x ** 0.5`? This expression should be both working for tensors and python scalars (does `x ** 0.5` torch.pow call dispatch to torch.sqrt directly if power == 0.5?)
